src/blr/cli/tagfastq.py

```diff
@@ -0,0 +1,132 @@
+"""
+Tag FASTQ/FASTA headers with corrected barcodes from separate raw barcode FASTQ with matching headers and starcode
+ CLSTR output file with corrected barcodes.
+"""
+
+import logging
+import sys
+import dnaio
+from tqdm import tqdm
+from collections import namedtuple
+
+logger = logging.getLogger(__name__)
+
+
+def main(args):
+    logger.info(f'Starting')
+
+    # Get the corrected barcodes and create a dictionary pointing each raw barcode to its canonical sequence.
+
+    with open(args.barcodes_corrected, "r") as reader:
+        corrected_barcodes = parse_corrected_barcodes(reader)
+
+    # Get the raw barcodes and create a dictionary pointing each read header to the raw barcode sequence.
+    with dnaio.open(args.barcodes_raw, mode="r") as reader:
+        header_barcodes = parser_raw_barcodes(reader)
+
+    in_interleaved = True if not args.input2 else False
+    logger.info(f"Input detected as {'interleaved fastq.' if in_interleaved else 'paired fastq.'}")
+
+    # If no output1 is given output is sent to stdout
+    if not args.output1:
+        logger.info(f"Writing output to stdout.")
+        args.output1 = sys.stdout.buffer
+        args.output2 = None
+
+    out_interleaved = True if not args.output2 else False
+    logger.info(f"Output detected as {'interleaved fastq.' if out_interleaved else 'paired fastq.'}")
+
+    reads_skipped = list()
```

Comment:
Acctually this will probably occur a bit as we have separated the trimming on the reads from the barcode extraction. Then are currently having some small differences in output that i think might persist. Thus I think it is good to log them somewhere, possibly we could also just write them to a separate file with unknown reads.

Reply:
>I think it would make more sense to keep a count on how many reads are without barcode, or if you really find it useful, at least make it a set rather than a list for memory efficiency.

FYI [sets are more memory consuming than lists. ](https://stackoverflow.com/questions/13547883/is-python-set-more-space-efficient-than-list)